operator/api/v1alpha2/seldondeployment_webhook.go

```diff
@@ -0,0 +1,446 @@
+					pu.Endpoint.ServiceHost = "localhost"
+				} else {
+					containerServiceValue := GetContainerServiceName(r, p, con)
+					pu.Endpoint.ServiceHost = containerServiceValue + "." + r.ObjectMeta.Namespace + ".svc.cluster.local."
+				}
+			}
+			if pu.Endpoint.ServicePort == 0 {
+				pu.Endpoint.ServicePort = portNum
+			}
+
+			// for prepack servers we want to add a container name and image to correspond to grafana dashboards
+			if IsPrepack(pu) {
+
+				existing := con != nil
+				if !existing {
+					con = &corev1.Container{
+						Name: pu.Name,
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								Name:      PODINFO_VOLUME_NAME,
+								MountPath: PODINFO_VOLUME_PATH,
+							},
+						},
+					}
+				}
+
+				SetImageNameForPrepackContainer(pu, con)
+
+				// if new Add container to componentSpecs
+				if !existing {
+					if len(p.ComponentSpecs) > 0 {
+						p.ComponentSpecs[0].Spec.Containers = append(p.ComponentSpecs[0].Spec.Containers, *con)
+					} else {
+						podSpec := SeldonPodSpec{
+							Metadata: metav1.ObjectMeta{CreationTimestamp: metav1.Now()},
+							Spec: corev1.PodSpec{
+								Containers: []corev1.Container{*con},
+							},
+						}
+						p.ComponentSpecs = []*SeldonPodSpec{&podSpec}
+
+						// p is a copy so update the entry
+						r.Spec.Predictors[i] = p
+					}
+				}
+			}
+
+		}
+
+	}
+
+}
+
+// -----
+
+// --- Validating
+
+// Check the predictive units to ensure the graph matches up with defined containers.
+func checkPredictiveUnits(pu *PredictiveUnit, p *PredictorSpec, fldPath *field.Path, allErrs field.ErrorList) field.ErrorList {
+	if *pu.Implementation == UNKNOWN_IMPLEMENTATION {
+
+		if GetContainerForPredictiveUnit(p, pu.Name) == nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, pu.Name, "Can't find container for Predictive Unit"))
+		}
+
+		if *pu.Type == UNKNOWN_TYPE && (pu.Methods == nil || len(*pu.Methods) == 0) {
+			allErrs = append(allErrs, field.Invalid(fldPath, pu.Name, "Predictive Unit has no implementation methods defined. Change to a know type or add what methods it defines"))
+		}
+
+	} else if *pu.Implementation == SKLEARN_SERVER ||
+		*pu.Implementation == XGBOOST_SERVER ||
+		*pu.Implementation == TENSORFLOW_SERVER ||
+		*pu.Implementation == MLFLOW_SERVER {
+		if pu.ModelURI == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath, pu.Name, "Predictive unit modelUri required when using standalone servers"))
+		}
+	}
+
+	for i := 0; i < len(pu.Children); i++ {
+		allErrs = checkPredictiveUnits(&pu.Children[i], p, fldPath.Index(i), allErrs)
+	}
+
+	return allErrs
+}
+
+func checkTraffic(mlDep *SeldonDeployment, fldPath *field.Path, allErrs field.ErrorList) field.ErrorList {
+	var trafficSum int32 = 0
+	for i := 0; i < len(mlDep.Spec.Predictors); i++ {
+		p := mlDep.Spec.Predictors[i]
+		trafficSum = trafficSum + p.Traffic
+	}
+	if trafficSum != 100 && len(mlDep.Spec.Predictors) > 1 {
```

Comment:
these two checks overlaps, and I am not sure if that is what you want.  this one should check for > 100 instead of != 100?

Reply:
Not sure they do. for multiple predictors your traffic must sum to 100. For 1 predictor traffic should not be set 0 or 100.